### ISSUE TITLE
Read current disk and output to disk-config.nix

### ISSUE THREAD

**Issue by FatBoyXPC** [nix-community/disko#728]
Is an option like this available? It would be pretty handy to be able to run a command that reads the current disk(s) specified and outputs to a `disk-config.nix`. One example of why this is handy is if I want to install nix into a laptop that's already dual booting windows and arch. I'd rather not blow away the windows partition as sometimes windows can be helpful in troubleshooting hardware.

**Comment by iFreilicht** [comment]
This does not currently exist, but it would be very handy indeed!

**Comment by Mic92** [comment]
Would be a good addition to have. Just now it's also possible to run disko with existing partitions, but we have not covered all edge cases yet. So there might be cases where it eats your data.

**Comment by iFreilicht** [comment]
I feel like a proper implementation of this would also open the door to safer idempotent `--mode format`. It's even kind-of required to make sure the implementations for serializing the current state and making only necessary changes don't drift apart.

Basically, my idea would be to try to serialize the current state of drives into a JSON first. From this, a Nix attrset can be generated trivially. That's what this feature is about.

As a follow-up, when running disko, it would do the same serialization and then diff that against the final attrset of the current configuration. Only the differences would then be taken into account when formatting.

Maybe this should even be a new mode.

As for how to implement this, I would say the `_create` attribute for each type should be joined by a `_serialize`. This doesn't work for the top level, but once we found the drives, the partitions on them and their types, we can dispatch to each of those to figure out the values of their options.

For the follow-up, a `_modify` attribute would probably be the way to go.

**Comment by Mic92** [comment]
Where would you store this json file?

**Comment by iFreilicht** [comment]
In a temporary directory or a fifo even. It should be regenerated for both the disko config and the real state on every run.

**Comment by Mic92** [comment]
Would it be easier at some point to just dump all the json and have a language that is more high-level than bash to use this as an input and compare it with reality? The error handling in bash is not very good and the templating at some stage also becomes not so readable.

**Comment by iFreilicht** [comment]
Yes, 100%. I feel like disko in general might be easier to maintain if instead of letting nix generate the final script, we "just" wrote a program that runs `nix eval -f disko.nix --json` and then works from those values, but that's a whole separate discussion, and there might be good reasons against it.

In terms of which language to use, do you or @Lassulus have a preference? If we're just talking about a single script, python seems to be the way to go, but if we want to expand the functionality further and actually make this a solid tool that uses as little bash as possible, Rust or Go could be good options as well.

**Comment by almereyda** [comment]
Please note that since [v2.3.0](https://github.com/openzfs/zfs/releases/tag/zfs-2.3.0) ZFS supports JSON output for most of its commands.
